src/calculate_bucket.hpp

```diff
@@ -154,69 +159,25 @@ class F1Calculator {
         return std::make_pair(CalculateF(L), L);
     }
 
-    // Returns an evaluation of F1(L), and the metadata (L) that must be stored to evaluate F2,
-    // for 'number_of_evaluations' adjacent inputs.
-    inline std::vector<std::pair<Bits, Bits>> CalculateBuckets(
-        const Bits& start_L,
-        uint64_t number_of_evaluations)
+    // F1(x) values for x in range [first_x, first_x + n) are placed in res[].
+    // n must not be more than 1 << kBatchSizes.
+    void CalculateBuckets(uint64_t first_x, uint64_t n, uint64_t *res)
     {
-        uint16_t num_output_bits = k_;
-        uint16_t block_size_bits = kF1BlockSizeBits;
+        uint64_t start = first_x * k_ / kF1BlockSizeBits;
+        uint64_t end = ((first_x + n + 1) * k_ - 1) / kF1BlockSizeBits;
+        uint64_t n_blks = end - start + 1;
+        uint32_t start_bit = first_x * k_ % kF1BlockSizeBits;
+        uint8_t x_shift = k_ - kExtraBits;
 
-        uint64_t two_to_the_k = (uint64_t)1 << k_;
-        if (start_L.GetValue() + number_of_evaluations > two_to_the_k) {
-            throw "Evaluation out of range";
-        }
-        // Counter for the first input
-        uint64_t counter = (start_L.GetValue() * (uint128_t)num_output_bits) / block_size_bits;
-        // Counter for the last input
-        uint64_t counter_end =
-            ((start_L.GetValue() + (uint128_t)number_of_evaluations + 1) * num_output_bits) /
-            block_size_bits;
-
-        std::vector<Bits> blocks;
-        uint64_t L = (counter * block_size_bits) / num_output_bits;
-        uint8_t ciphertext_bytes[kF1BlockSizeBits / 8];
+        chacha8_get_keystream(&this->enc_ctx_, start, n_blks, buf_);
+        for (uint64_t i = 0; i < n; i++) {
+            uint64_t x = first_x + i;
+            uint64_t y = Util::SliceInt64FromBytes(buf_, start_bit, k_);
 
-        // Evaluates ChaCha8 for each block
-        while (counter <= counter_end) {
-            chacha8_get_keystream(&this->enc_ctx_, counter, 1, ciphertext_bytes);
-            Bits ciphertext(ciphertext_bytes, block_size_bits / 8, block_size_bits);
-            blocks.push_back(std::move(ciphertext));
-            ++counter;
-        }
-
-        std::vector<std::pair<Bits, Bits>> results;
-        uint64_t block_number = 0;
-        uint16_t start_bit = (start_L.GetValue() * (uint128_t)num_output_bits) % block_size_bits;
+            res[i] = (y << kExtraBits) | (x >> x_shift);
```

Comment:
there is an assumption that `_k + kExtraBits <= 64`, right? Otherwise bits from `y` would be lost.

Reply:
Yes, that's correct. Minimum and maximum `k_` values are defined by `kMinPlotSize` and `kMaxPlotSize`, which are 18 and 50. `kExtraBits` is 6.